StreamlinersNZ/interview-nodejs
# Adds dominant color extraction feature for images

**Task Description**
 - [x] Forked the repository.
 - [x] Added a postman end point to accept jpg format image.
 - [x] Added the function named `extractDominantColor`  to process and determine dominant colour of the image.
 - [x] Returned the RGB values of the colour through the endpoint.
 - [x] Added Tests to check the functionality of the function named `extractDominantColor`.
 
**What kind of change does this PR introduce?**
- **Feature request** - Add dominant colour extractor.

**What is the current behaviour ?**
- Currently the application could not extract dominant colour from the uploaded image.

**How to test the new behaviour?**
1) Run the server using command `npm run start` , the server will start listening on port `3000`.
2) Create the POST end point using *postman* through `http://localhost:3000/api/dominantcolor/upload`.
3) Add to the body of the POST request, with following parameter: 
                 key named "image", and value with the image file. 

![image](https://github.com/StreamlinersNZ/interview-nodejs/assets/54251621/ea8d608d-ef37-412a-ac4e-c55d81650a50)

4) Send the request through *postman*.
5) Verify the response  (example below)

![image](https://github.com/StreamlinersNZ/interview-nodejs/assets/54251621/7e634592-7c39-48a5-9ed3-9e658dc5c5b3)

 **Other Information**
- The chosen library *Vibrant* gave best results
- Other libraries like *JIMP* etc were explored but gave conflicting results for known cases.
- A *Custom Jest Matcher*  is written to compare the array values 
- A sample image file is added to the assets folder.
